product_internal_reference_generator/views/product.xml

```diff
@@ -0,0 +1,129 @@
+<?xml version='1.0' encoding='utf-8' ?>
+<odoo>
+    <record model="ir.ui.view" id="product_template_form_view_inherit">
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="product.product_template_only_form_view" />
+        <field name="arch" type="xml">
+            <field name="default_code" position="before">
+                <label
+                    for="int_ref_template_id"
+                    class="o_form_label"
+                    attrs="{'invisible': [('variants_prefix', '!=', False)]}"
+                    string="Int. Ref. Template"
+                />
+                <div
+                    class="o_row"
+                    name="sequence"
+                    attrs="{'invisible': [('variants_prefix', '!=', False)]}"
+                >
+                    <field name="int_ref_template_id" />
+                    <button
+                        name="btn_generate_sequence"
+                        type="object"
+                        string="Generate"
+                        attrs="{'invisible': [('int_ref_template_id', '=', False)]}"
+                        confirm="Are you sure want to generate new reference prefix?
+                        If variants already exist for this product, the new internal
+                        reference will be assigned only those where internal
+                        reference is not set."
+                    />
+                </div>
+                <field
+                    name="int_ref_template_id"
+                    string="Int. Ref. Template"
+                    attrs="{'invisible': [('variants_prefix', '=', False)]}"
+                    groups="product_internal_reference_generator.group_int_ref_template_always_visible"
+                />
+                <field
+                    name="variants_prefix"
+                    attrs="{'invisible': [('int_ref_template_id', '=', False)]}"
+                    string="Int. Ref. Prefix"
+                />
+                <field name="default_code_editable" invisible="1" />
+            </field>
+            <field name="default_code" position="attributes">
+                <attribute name="attrs">{
+                    'readonly': [('default_code_editable', '!=', True)],
+                    'invisible': [('product_variant_count', '>', 1)]
+                }</attribute>
```

Comment:
**suggestion**: consider adding a dependency to https://github.com/OCA/server-tools/blob/14.0/base_view_inheritance_extension so you can only add the `readonly` key and not touch the `invisible` key. This would improve the compatibility of this module with other modules that also change the `attrs` of`default_code`